akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala

```diff
@@ -1387,6 +1450,47 @@ final class Flow[In, Out, Mat](delegate: scaladsl.Flow[In, Out, Mat]) extends Gr
   def recoverWithRetries(attempts: Int, pf: PartialFunction[Throwable, Graph[SourceShape[Out], NotUsed]]): javadsl.Flow[In, Out, Mat] =
     new Flow(delegate.recoverWithRetries(attempts, pf))
 
+  /**
+    * RecoverWithRetries allows to switch to alternative Source on flow failure. It will stay in effect after
+    * a failure has been recovered up to `attempts` number of times so that each time there is a failure
+    * it is fed into the `pf` and a new Source may be materialized. Note that if you pass in 0, this won't
+    * attempt to recover at all.
+    *
+    * A negative `attempts` number is interpreted as "infinite", which results in the exact same behavior as `recoverWith`.
+    *
+    * Since the underlying failure signal onError arrives out-of-band, it might jump over existing elements.
+    * This stage can recover the failure signal, but not the skipped elements, which will be dropped.
+    *
+    * Throwing an exception inside `recoverWithRetries` _will_ be logged on ERROR level automatically.
+    *
+    * '''Emits when''' element is available from the upstream or upstream is failed and element is available
+    * from alternative Source
+    *
+    * '''Backpressures when''' downstream backpressures
+    *
+    * '''Completes when''' upstream completes or upstream failed with exception pf can handle
+    *
+    * '''Cancels when''' downstream cancels
+    *
+    * @param attempts Maximum number of retries or -1 to retry indefinitely
+    * @param clazz the class object of the failure cause
+    * @param graph the new Source to be materialized
+    */
+  def recoverWithRetries(attempts: Int, clazz: Class[_ <: Throwable], graph: Graph[SourceShape[Out], NotUsed]): javadsl.Flow[In, Out, Mat] = {
+    import JavaPartialFunction._
+
+    val pf = new JavaPartialFunction[Throwable, Graph[SourceShape[Out], NotUsed]] {
+      override def apply(x: Throwable, isCheck: Boolean): Graph[SourceShape[Out], NotUsed] =
+        if (x.getClass == clazz) {
+          if (isCheck) null
+          else graph
+        }
+        else throw noMatch()
+    }
+
+    recoverWithRetries(attempts, pf)
+  }
```

Comment:
Probably easier to implement this as:
```
new Flow(delegate.recoverWithRetries(attempts, { case elem if clazz.isInstance(elem)} => elem))
```

Reply:
Oh, yes, this is a more elegant implementation, thanks 👍